### PR TITLE
Fixing labels not appearing when enabling terrain at high zoom

### LINKED ISSUE #3454: Labels don't appear when enabling terrain at high zooms
**maplibre-gl-js version**: main

**browser**: Firefox and Chrome

### Steps to Trigger Behavior

 1. Create a map with labels and a terrain source ([e.g. MapTiler's example](https://www.maptiler.com/news/2022/05/maplibre-v2-add-3d-terrain-to-your-map/))
 2. Ensure terrain is disabled
 3. Zoom in (>15.5)
 4. Enable terrain

### Link to Demonstration

https://codepen.io/snailbones/pen/rNPoZpe

### Expected Behavior
Labels should be visible. Here is the 2d map before enabling terrain:
image

### Actual Behavior

Labels are missing. They do not load until the user triggers a map update.
image

### PR DESCRIPTION
**Pull request by SnailBones**
Fixes #3454

## Before:
[broken.jpg](https://github.com/maplibre/maplibre-gl-js/assets/14878684/b2590d5b-617f-40b2-8d89-d32c09c69fa6)

## After:
[fixed.webm](https://github.com/maplibre/maplibre-gl-js/assets/14878684/69f94f64-fedc-484d-bfdd-49b7cfd49d80)

## Launch Checklist

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
If I understood the fix correctly, the fix is done by reloading the source cache. Can you clarify why reloading the terrain source cache solves this issue?

File: test/integration/render/tests/terrain/symbol-add-terrain/style.json [hunk 1]
```diff
@@ -0,0 +1,117 @@
+{
+    "version": 8,
+    "sprite": "local://sprites/sprite",
+    "glyphs": "local://glyphs/{fontstack}/{range}.pbf",
+    "metadata": {
+        "test": {
+            "height": 512,
+            "width": 512,
+            "operations": [
+                [
```
Comment (HarelM) [inline test/integration/render/tests/terrain/symbol-add-terrain/style.json:10]: Is there a way to reduce the time this test will run? Maybe using idle event?

**Comment by SnailBones** [comment]
> If I understood the fix correctly, the fix is done by reloading the source cache. Can you clarify why reloading the terrain source cache solves this issue?

My understanding of this error is that it's caused by the terrain source cache remaining in an incorrect state when terrain is loaded during rendering. My fix is to first set `painter.terrainFacilitator.dirty = true;` so that labels will be redrawn, and then reload tiles in the terrain source cache, which triggers another render loop after terrain is correctly loaded.

File: test/integration/render/tests/terrain/symbol-add-terrain/style.json (hunk 1 above)
Comment (SnailBones) [reply test/integration/render/tests/terrain/symbol-add-terrain/style.json:11]: Good call! Updated it to remove the timed waits.

**Review by HarelM — APPROVED** [review]
THANKS!

**Pull request merged**